trax/models/research/funnel_transformer.py

```diff
@@ -821,21 +828,148 @@ def create_reformer_blocks(n_layers, dense=True):
   post_decoder_blocks = create_reformer_blocks(n_post_decoder_blocks,
                                                dense=False)
 
+  cacher = RelformerCacher(total_kv_pooling=shorten_factor,
+                           n_raw_tokens_generated=n_raw_tokens_generated,
+                           max_inference_length=max_len,
+                           shift=shorten_factor - 1,
+                           mode=mode)
+
+  picker = RelformerPicker(total_kv_pooling=shorten_factor,
+                           n_raw_tokens_generated=n_raw_tokens_generated,
+                           mode=mode)
+
+  cacher_conv = RelformerCacher(total_kv_pooling=shorten_factor,
+                                n_raw_tokens_generated=n_raw_tokens_generated,
+                                max_inference_length=max_len,
+                                shift=shorten_factor - 1,
+                                sliding=True,
+                                mode=mode)
+
+  picker_conv = PickLastTokenInPredict(mode=mode)
+
   # Assemble and return the model.
   return tl.Serial(              # tokens (or chunked tuple of tokens)
       tl.ShiftRight(mode=mode),  # toks
       token_encoder,             # vecs
       positional_encoder,
       pre_decoder_blocks,        # vecs
       tl.Dup(),
-      tl.ShiftRight(n_positions=shorten_factor - 1),
+      cacher,
+      tl.ShiftRight(n_positions=shorten_factor - 1, mode=mode),
       _DownsamplerLM(shorten_factor, d_model),
       relative_decoder_blocks,
       tl.Dropout(rate=dropout, shared_axes=[-2], mode=mode),
       _UpsamplerLM(shorten_factor, d_model),
       tl.LayerNorm(),
+      picker,
       tl.Concatenate(),
+      cacher_conv,
       conv_layer,
+      picker_conv,
       post_decoder_blocks,
-      tl.Dense(vocab_size),      # vecs
+      tl.Dense(vocab_size),  # vecs
   )
+
+
+class RelformerCacher(tl.Layer):
```

Comment:
I would suggest moving these new layers above so that they are between FunnelTransformerLM and RelformerLM (this is default convention in this file to have helper layers above the model itself).